File: sdk/ios/Sources/Vdb/Filter.swift
import CVdb
import Foundation

/// A metadata predicate.
///
/// A tree, because that is what a filter is. The C ABI receives it as a postfix sequence — the
/// only shape C can take a tree in without thirty functions — and this type is what keeps that
/// out of sight: callers write an expression and it is flattened on the way down.
///
/// Evaluation is **total**. Comparing a string to a number is `false`, never an error; a field
/// no document has is absent. Three consequences surprise people, and each is deliberate:
///
/// - An absent field equals `.null`, so `.equals("x", .null)` matches a document without `x`.
///   Use ``exists(_:)`` when the distinction matters.
/// - ``notEquals(_:_:)`` is the exact negation of ``equals(_:_:)``, so it matches absent fields.
/// - ``greaterThan(_:_:)`` and ``lessThanOrEqual(_:_:)`` are *both* false where no ordering is
///   defined, so they are not negations of one another.
///
/// The full rules are in `docs/api/filters.md`.
public indirect enum Filter: Sendable {
    /// A value a filter can compare against.
    public enum Value: Sendable, Equatable {
        case string(String)
        case int(Int64)
        case double(Double)
        case bool(Bool)
    }

    case equals(String, Value)
    case notEquals(String, Value)
    case greaterThan(String, Value)
    case greaterThanOrEqual(String, Value)
    case lessThan(String, Value)
    case lessThanOrEqual(String, Value)
    /// The field is a string with this prefix.
    case startsWith(String, String)
    /// The field is an **array** containing this value. Not a substring test.
    case contains(String, Value)
    /// The field is present, including an explicit null.
    case exists(String)
    /// The field is absent, or present and null.
    case isNull(String)

    /// Every child must match. An empty `all` matches everything.
    case all([Filter])
    /// At least one child must match. An empty `any` matches nothing.
    case any([Filter])
    case not(Filter)

    /// Combine two filters, requiring both.
    public static func && (lhs: Filter, rhs: Filter) -> Filter { .all([lhs, rhs]) }
    /// Combine two filters, requiring either.
    public static func || (lhs: Filter, rhs: Filter) -> Filter { .any([lhs, rhs]) }
    /// Negate a filter.
    public static prefix func ! (operand: Filter) -> Filter { .not(operand) }
}

extension Filter {
    /// Flatten into the ABI's builder, in postfix order.
    ///
    /// A post-order walk: children first, then the combinator that consumes them. The builder's
    /// depth invariant — one expression left at the end — falls out of the traversal, so an
    /// unbalanced sequence is not something a caller of this API can construct.
    func encode(into builder: OpaquePointer) throws {
        switch self {
        case let .equals(field, value): try push(builder, field, VDB_OP_EQ, value)
        case let .notEquals(field, value): try push(builder, field, VDB_OP_NE, value)
        case let .greaterThan(field, value): try push(builder, field, VDB_OP_GT, value)
        case let .greaterThanOrEqual(field, value): try push(builder, field, VDB_OP_GTE, value)
        case let .lessThan(field, value): try push(builder, field, VDB_OP_LT, value)
        case let .lessThanOrEqual(field, value): try push(builder, field, VDB_OP_LTE, value)
        case let .startsWith(field, prefix):
            try push(builder, field, VDB_OP_STARTS_WITH, .string(prefix))
        case let .contains(field, value): try push(builder, field, VDB_OP_CONTAINS, value)

        case let .exists(field): try pushUnary(builder, field, VDB_UNARY_EXISTS)
        case let .isNull(field): try pushUnary(builder, field, VDB_UNARY_IS_NULL)

        case let .all(children):
            // An empty `all` matches everything, which the engine spells as an empty
            // conjunction — but the ABI refuses a zero-count combine, since at that level it is
            // far more likely to be a mistake. Expressed instead as a tautology.
            if children.isEmpty {
                try pushUnary(builder, "", VDB_UNARY_IS_NULL)
                try pushUnary(builder, "", VDB_UNARY_IS_NULL)
                try combine(builder, VDB_COMBINE_OR, 2)
                return
            }
            for child in children { try child.encode(into: builder) }
            try combine(builder, VDB_COMBINE_AND, children.count)

        case let .any(children):
            if children.isEmpty {
                // An empty `any` matches nothing: a field is both null and present.
                try pushUnary(builder, "", VDB_UNARY_IS_NULL)
                try pushUnary(builder, "", VDB_UNARY_EXISTS)
                try combine(builder, VDB_COMBINE_AND, 2)
                return
            }
            for child in children { try child.encode(into: builder) }
            try combine(builder, VDB_COMBINE_OR, children.count)

        case let .not(child):
            try child.encode(into: builder)
            try combine(builder, VDB_COMBINE_NOT, 1)
        }
    }

    private func push(
        _ builder: OpaquePointer, _ field: String, _ op: vdb_op_t, _ value: Value
    ) throws {
        var fieldBytes = Array(field.utf8)
        let rawOp = Int32(op.rawValue)
        try check { error in
            fieldBytes.withUnsafeMutableBufferPointer { f in
                switch value {
                case let .string(s):
                    var bytes = Array(s.utf8)
                    return bytes.withUnsafeMutableBufferPointer { v in
                        vdb_filter_compare_str(
                            builder, f.baseAddress, f.count, rawOp, v.baseAddress, v.count, error
                        )
                    }
                case let .int(i):
                    return vdb_filter_compare_i64(builder, f.baseAddress, f.count, rawOp, i, error)
                case let .double(d):
                    return vdb_filter_compare_f64(builder, f.baseAddress, f.count, rawOp, d, error)
                case let .bool(b):
                    return vdb_filter_compare_bool(builder, f.baseAddress, f.count, rawOp, b, error)
                }
            }
        }
    }

    private func pushUnary(
        _ builder: OpaquePointer, _ field: String, _ predicate: vdb_unary_t
    ) throws {
        var bytes = Array(field.utf8)
        let raw = Int32(predicate.rawValue)
        try check { error in
            bytes.withUnsafeMutableBufferPointer { f in
                vdb_filter_unary(builder, f.baseAddress, f.count, raw, error)
            }
        }
    }

    private func combine(
        _ builder: OpaquePointer, _ combinator: vdb_combine_t, _ count: Int
    ) throws {
        let raw = Int32(combinator.rawValue)
        try check { vdb_filter_combine(builder, raw, count, $0) }
    }
}

File: sdk/android/src/main/java/dev/vdb/Filter.java
package dev.vdb;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * A metadata predicate.
 *
 * <p>A tree, because that is what a filter is. The native layer receives it as a postfix
 * sequence — the only shape JNI can take a tree in without reflection on every node — and this
 * class is what keeps that out of sight: callers write an expression and it is flattened on the
 * way down, so an unbalanced sequence is not something they can construct.
 *
 * <pre>{@code
 * Filter cheapTools = Filter.eq("category", "tools").and(Filter.lt("price", 50.0));
 * List<Collection.Hit> hits = docs.search(query, 10, cheapTools);
 * }</pre>
 *
 * <p>Evaluation is <b>total</b>: comparing a string to a number is {@code false}, never an
 * error, and a field no document has is absent. Three consequences surprise people, and each is
 * deliberate — an absent field equals null; {@link #ne} is the exact negation of {@link #eq} so
 * it matches absent fields; and {@link #gt} and {@link #lte} are <em>both</em> false where no
 * ordering is defined, so they are not negations of one another. See {@code docs/api/filters.md}.
 *
 * <p>Plain classes rather than sealed interfaces or records: this ships to Android, where those
 * need a recent API level or desugaring, and a filter type is not worth that constraint.
 */
public abstract class Filter {
  // Operators, mirroring the constants on the Rust side.
  private static final int OP_EQ = 1;
  private static final int OP_NE = 2;
  private static final int OP_GT = 3;
  private static final int OP_GTE = 4;
  private static final int OP_LT = 5;
  private static final int OP_LTE = 6;
  private static final int OP_STARTS_WITH = 7;
  private static final int OP_CONTAINS = 8;

  private static final int UNARY_EXISTS = 1;
  private static final int UNARY_IS_NULL = 2;

  private static final int COMBINE_AND = 1;
  private static final int COMBINE_OR = 2;
  private static final int COMBINE_NOT = 3;

  Filter() {}

  /** Flatten this node onto a native builder, children first. */
  abstract void encode(long builder);

  // ---- leaves ----

  public static Filter eq(String field, String value) {
    return new Compare(field, OP_EQ, Value.of(value));
  }

  public static Filter eq(String field, long value) {
    return new Compare(field, OP_EQ, Value.of(value));
  }

  public static Filter eq(String field, double value) {
    return new Compare(field, OP_EQ, Value.of(value));
  }

  public static Filter eq(String field, boolean value) {
    return new Compare(field, OP_EQ, Value.of(value));
  }

  /** The exact negation of {@link #eq}, so it matches documents lacking the field. */
  public static Filter ne(String field, String value) {
    return new Compare(field, OP_NE, Value.of(value));
  }

  public static Filter ne(String field, long value) {
    return new Compare(field, OP_NE, Value.of(value));
  }

  public static Filter gt(String field, double value) {
    return new Compare(field, OP_GT, Value.of(value));
  }

  public static Filter gte(String field, double value) {
    return new Compare(field, OP_GTE, Value.of(value));
  }

  public static Filter lt(String field, double value) {
    return new Compare(field, OP_LT, Value.of(value));
  }

  public static Filter lte(String field, double value) {
    return new Compare(field, OP_LTE, Value.of(value));
  }

  /** The field is a string with this prefix. */
  public static Filter startsWith(String field, String prefix) {
    return new Compare(field, OP_STARTS_WITH, Value.of(prefix));
  }

  /** The field is an <b>array</b> containing this value. Not a substring test. */
  public static Filter contains(String field, String value) {
    return new Compare(field, OP_CONTAINS, Value.of(value));
  }

  /** The field is present, including an explicit null. */
  public static Filter exists(String field) {
    return new Unary(field, UNARY_EXISTS);
  }

  /** The field is absent, or present and null. */
  public static Filter isNull(String field) {
    return new Unary(field, UNARY_IS_NULL);
  }

  // ---- combinators ----

  /** Every child must match. */
  public static Filter all(Filter... children) {
    return new Combine(COMBINE_AND, Arrays.asList(children));
  }

  /** At least one child must match. */
  public static Filter any(Filter... children) {
    return new Combine(COMBINE_OR, Arrays.asList(children));
  }

  public static Filter not(Filter child) {
    return new Combine(COMBINE_NOT, java.util.Collections.singletonList(child));
  }

  public Filter and(Filter other) {
    return all(this, other);
  }

  public Filter or(Filter other) {
    return any(this, other);
  }

  // ---- flattening ----

  /** Build a native filter handle. The caller must free it. */
  long toNative() {
    long handle = Native.filterNew();
    try {
      encode(handle);
      int depth = Native.filterDepth(handle);
      if (depth != 1) {
        // Unreachable through this API — the traversal guarantees it — but a wrong answer here
        // would be a filter that silently drops a clause, so it is checked rather than trusted.
        throw new VdbException("filter flattened to depth " + depth + ", expected 1");
      }
    } catch (RuntimeException e) {
      Native.filterFree(handle);
      throw e;
    }
    return handle;
  }

  private static final class Compare extends Filter {
    private final String field;
    private final int op;
    private final Value value;

    Compare(String field, int op, Value value) {
      this.field = field;
      this.op = op;
      this.value = value;
    }

    @Override
    void encode(long builder) {
      value.compareOn(builder, field, op);
    }
  }

  private static final class Unary extends Filter {
    private final String field;
    private final int predicate;

    Unary(String field, int predicate) {
      this.field = field;
      this.predicate = predicate;
    }

    @Override
    void encode(long builder) {
      Native.filterUnary(builder, field, predicate);
    }
  }

  private static final class Combine extends Filter {
    private final int combinator;
    private final List<Filter> children;

    Combine(int combinator, List<Filter> children) {
      this.combinator = combinator;
      this.children = new ArrayList<>(children);
    }

    @Override
    void encode(long builder) {
      if (children.isEmpty()) {
        // The identities: an empty `all` matches everything, an empty `any` matches nothing.
        // The native builder refuses a zero-count combine, where it is far more likely a
        // mistake, so these are written as an explicit tautology and contradiction.
        Native.filterUnary(builder, "", UNARY_IS_NULL);
        if (combinator == COMBINE_AND) {
          Native.filterUnary(builder, "", UNARY_IS_NULL);
          Native.filterCombine(builder, COMBINE_OR, 2);
        } else {
          Native.filterUnary(builder, "", UNARY_EXISTS);
          Native.filterCombine(builder, COMBINE_AND, 2);
        }
        return;
      }
      for (Filter child : children) {
        child.encode(builder);
      }
      Native.filterCombine(builder, combinator, children.size());
    }
  }
}
